src/groupeddataframe/splitapplycombine.jl

```diff
@@ -168,63 +168,48 @@ function groupby(df::AbstractDataFrame, cols;
-- A single value gives a single additional column and one row per group.
-- A named tuple of single values or a [`DataFrameRow`](@ref) gives one additional column
-  for each field and one row per group.
-- A vector gives a single additional column and as many rows
+- A single value gives a `DataFrame` with a single additional column and one row per group.
+- A named tuple of single values or a [`DataFrameRow`](@ref) gives a `DataFrame` with
+  one additional column for each field and one row per group.
+- A vector gives a data frame with a single additional column and as many rows
   for each group as the length of the returned vector for that group.
-- A data frame, a named tuple of vectors or a matrix gives the same additional columns
-  and as many rows for each group as the rows returned for that group.
+- A data frame, a named tuple of vectors or a matrix gives a `DataFrame` with the same
+  additional columns and as many rows for each group as the rows returned for that group.
   As a special case, returning an empty table with zero columns is allowed,
   whatever the number of columns returned for other groups.
 
-`f` must always return the same kind of object (as defined in the above list) for
-all groups, and if a named tuple or data frame, with the same fields or columns.
-Named tuples cannot mix single values and vectors.
+`f` must always return the same kind of object (a single value or a table as defined
+in the above list) for all groups, and with the same column names (defined explicitly
+or auto-generated). If an empty vector, data frame, or named tuple is returned then
+a given group is dropped from the result.
+In particular, named tuples cannot mix single values and vectors.
 Due to type instability, returning a single value or a named tuple is dramatically
 faster than returning a data frame.
 
-As a special case, if a tuple or vector of pairs is passed as the first argument, each function
-is required to return a single value or vector, which will produce each a separate column.
-
-Column names are automatically generated when necessary: for functions
-operating on a single column and returning a single value or vector, the function name is
-appended to the input column name; for other functions, columns are called `x1`, `x2`
-and so on.
-
 Optimized methods are used when standard summary functions (`sum`, `prod`,
 `minimum`, `maximum`, `mean`, `var`, `std`, `first`, `last` and `length`)
-are specified using the pair syntax (e.g. `col => sum`).
+are specified using the `Pair` syntax (e.g. `col => sum`).
 When computing the `sum` or `mean` over floating point columns, results will be less
 accurate than the standard [`sum`](@ref) function (which uses pairwise summation). Use
 `col => x -> sum(x)` to avoid the optimized method and use the slower, more accurate one.
 
+Column names are automatically generated when necessary using the names `x1`, `x2` and so on
+if `f` is a `Function` or an `AbstractMatrix` is returned by it
```

Comment:
"or" should be "or". We use names x1 etc. if we use `Function` interface OR if we use `Pair` interface, but return an `AbstractMatrix`, otherwise in `Pair` interface for `AbstractVector` automatic column generation does not produce x1 as in the past but nice names like `select` does. I will try to explain it more.

The relevant part in `select` is:
```
Columns can be renamed using the `old_column => new_column_name` syntax,
and transformed using the `old_column => fun => new_column_name` syntax.
`new_column_name` must be a `Symbol`, and `fun` a function or a type. If `old_column`
is a `Symbol` or an integer then `fun` is applied to the corresponding column vector.
Otherwise `old_column` can be any column indexing syntax, in which case `fun`
will be passed the column vectors specified by `old_column` as separate arguments.
If `fun` returns a value of type other than `AbstractVector` then it will be wrapped
into a 1-element vector, unless its type is one of `AbstractDataFrame`, `NamedTuple`,
`DataFrameRow`, `AbstractMatrix`, in which case an error is thrown as currently these
return types are not allowed.

To apply `fun` to each row instead of whole columns, it can be wrapped in a `ByRow`
struct. In this case if `old_column` is a `Symbol` or an integer then `fun` is applied
to each element (row) of `old_column` using broadcasting. Otherwise `old_column` can be
any column indexing syntax, in which case `fun` will be passed one argument for each of
the columns specified by `old_column`. If `ByRow` is used it is not allowed for
`old_column` to select an empty set of columns.

Column transformation can also be specified using the short `old_column => fun` form.
In this case, `new_column_name` is automatically generated as `\$(old_column)_\$(fun)`.
Up to three column names are used for multiple input columns and they are joined
using `_`; if more than three columns are passed then the name consists of the
first two names and `etc` suffix then, e.g. `[:a,:b,:c,:d] => fun` produces
the new column name `:a_b_etc_fun`.
```
I thought it would be too long (probably some minor update could be done). What do you think?

Reply:
> "or" should be "or". We use names x1 etc. if we use `Function` interface OR if we use `Pair` interface, but return an `AbstractMatrix`, otherwise in `Pair` interface for `AbstractVector` automatic column generation does not produce x1 as in the past but nice names like `select` does. I will try to explain it more.

OK, got it. I think I had missed "when necessary". Though it's quite hard to parse since you need to combine lots of conditions. Maybe it would be simpler to say in which cases the `select` rules apply to generate a nice name, and that `x1`, etc. are used otherwise? Just wondering.

(Otherwise, I agree the `select` docstring is too verbose.)